substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrChunkWriter.java

```diff
@@ -449,6 +584,87 @@ private void processSamplerBuffers0() {
+        traverseList(getNativeBufferList(), false, false);
+
+        JfrBuffers buffers = globalMemory.getBuffers();
+        for (int i = 0; i < globalMemory.getBufferCount(); i++) {
+            JfrBuffer buffer = buffers.addressOf(i).read();
+            if (!JfrBufferAccess.acquire(buffer)) { // one attempt
+                continue;
+            }
+            write(buffer);
+            JfrBufferAccess.reinitialize(buffer);
+            JfrBufferAccess.release(buffer);
+        }
+    }
+
+    @Uninterruptible(reason = "Called from uninterruptible code.")
+    private void traverseList(JfrBufferNodeLinkedList linkedList, boolean java, boolean safepoint) {
+        // Traverse back to front to minimize conflict with threads adding new nodes.
+        // Which could possibly block traversal early at the head.
+
+        JfrBufferNode node = linkedList.getAndLockTail();
+        // If we couldn't acquire the tail. node is null. Give up, and try again next time.
+
+        while (node.isNonNull()) {
+            boolean writeSuccess = false;
+            VMError.guarantee(isAcquired(node) || VMOperation.isInProgressAtSafepoint(), "Cannot traverse JfrBufferNodeLinkedList outside safepoint without acquiring nodes.");
+            JfrBufferNode prev = node.getPrev();
+
+            if (!linkedList.isTail(node)) {
+                JfrBuffer buffer = node.getValue();
+                VMError.guarantee(buffer.isNonNull(), "JFR buffer should exist if we have not already removed its respective node.");
+
+                // Try to get BUFFER if not in safepoint
+                // make one attempt
+                if (!safepoint && !JfrBufferAccess.acquire(buffer)) {
+                    if (tryAcquire(prev)) {
+                        release(node);
+                        node = prev;
+                        continue;
+                    }
+                    release(node);
+                    return;
+                }
+                write(buffer);
+                writeSuccess = true;
+                if (!safepoint) {
+                    JfrBufferAccess.release(buffer);
+                }
+            }
+
+            // If a node has been marked dead, the thread was unable to flush upon death.
+            // So we need to flush the buffer above before attempting to remove the node here.
+            if (!node.getAlive()) {
+                if (linkedList.removeNode(node, true)) {
+                    // able to remove node
+                    if (tryAcquire(prev)) {
+                        node = prev;
+                        continue;
+                    }
+                    // Failed to get next node. Give up and try again later
+                    return;
+                }
+            } else if (writeSuccess && java) {
+                // Only notify event writer of flush if the thread (and its locals) is still alive
+                JfrThreadLocal.notifyEventWriter(node.getThread());
```

Comment:
oh that's true. I have removed this now. Although this would not have caused issues with the old code (because it was done within a safepoint) why was it necessary to do it then? We were still only flushing the committed data.

Reply:
Ok I see. So the event writer only needs to be notified only at a safepoint. Additionally, its only safe to notify at a safepoint anyway because if a node isn't marked retired, the corresponding thread must still be alive.